internal/core/application/trade_service.go

```diff
@@ -429,37 +424,69 @@ func (t *tradeService) tradeComplete(
 ) (txID string, swapFail domain.SwapFail, err error) {
 	trade, err := t.tradeRepository.GetTradeBySwapAcceptID(ctx, swapComplete.GetAcceptId())
 	if err != nil {
-		return "", nil, err
+		return
 	}
 
-	tradeID := trade.ID
-	err = t.tradeRepository.UpdateTrade(
-		ctx,
-		&tradeID,
-		func(trade *domain.Trade) (*domain.Trade, error) {
-			psetBase64 := swapComplete.GetTransaction()
-			res, err := trade.Complete(psetBase64)
-			if err != nil {
-				return nil, err
-			}
-			if !res.OK {
-				swapFail = trade.SwapFailMessage()
-				return trade, nil
-			}
+	psetBase64 := swapComplete.GetTransaction()
 
-			log.Info("trade with id ", tradeID, " completed")
+	// here we manipulate the trade to reach the Complete status
+	res, err := trade.Complete(psetBase64)
+	if err != nil {
+		return
+	}
+	// for domain related errors, we check for swap failures that can happens
+	// for tradin related problems or transaction manomission
+	if !res.OK {
+		swapFail = trade.SwapFailMessage()
+		return
+	}
+	log.Infof("trade with id %s completed", trade.ID)
 
-			if _, err := t.explorerSvc.BroadcastTransaction(res.TxHex); err != nil {
-				return nil, err
-			}
+	// we are going to broadcast the transaction, this will actually tell if the
+	//transaction is a valid one to be included in blockcchain
+	if _, err = t.explorerSvc.BroadcastTransaction(res.TxHex); err != nil {
+		return
+	}
 
-			txID = res.TxID
+	txID = res.TxID
+	log.Infof("trade with id %s broadcasted: %s", trade.ID, txID)
+
+	// we make sure that any problem happening at this point
+	// is not influencing the trade therefore we run as goroutine
+	// this method will take care to retry to handle potential
+	// datastore conflicts (if any) at repository level
+	go func() {
+		if err := t.tradeRepository.UpdateTrade(
+			ctx,
+			&trade.ID,
+			func(previousTrade *domain.Trade) (*domain.Trade, error) { return trade, nil },
+		); err != nil {
+			log.Error("unable to persist completed trade with id ", trade.ID, " : ", err.Error())
+		}
 
-			log.Info("trade with id ", tradeID, " broadcasted: ", txID)
+		_, accountIndex, _ := t.marketRepository.GetMarketByAsset(
+			ctx,
+			trade.MarketQuoteAsset,
+		)
+
+		unspentsToAdd, unspentsToSpend, err := extractUnspentsFromTx(
```

Comment:
@altafan can we use here extractUnspentsFromTxAndUpdateUtxoSet method?

Reply:
done